### PR TITLE
Adds one-shot mDNS resolution

### PR DESCRIPTION
**Pull request by benbrittain**
RFC 6762 Section 5.1 specifies a one-shot multicast DNS query. This
query has minimal differences from a standard DNS query, mostly just
using a multicast address and a different port (5353 vs 53).

A fully standards compliant mDNS implementation would use UDP source
port 5353 as well to issue queries, however we MUST NOT use that port
and continue using an ephemeral port until features such as service
discovery are implemented.

This change also allows specifying what kind of DNS query we wish to
perform.

https://www.rfc-editor.org/rfc/rfc6762#section-5.1

### CODE REVIEW

File: src/socket/dns.rs [hunk 1]
```diff
@@ -234,19 +251,33 @@ impl<'a> Socket<'a> {
...
         raw_name: &[u8],
+        query_type: Type,
+        mdns: bool,
     ) -> Result<QueryHandle, StartQueryError> {
         let handle = self.find_free_query().ok_or(StartQueryError::NoFreeSlot)?;
 
+        if mdns {
+            // as per RFC 6762 any DNS query ending in .local. MUST be sent as mdns
+            // so we internally overwrite the servers for any of those queries
+            self.update_servers(&[
```
Comment (timokroeger) [inline src/socket/dns.rs]: It looks like this line overrides the dns servers permanently. Which would mean that all following regular DNS queries will fail.

Comment (benbrittain) [reply src/socket/dns.rs]: Totally does, fixed.

**Comment by benbrittain** [comment]
@thvdveld Is this the sort of change you review as well? It's relevant in a 6lowpan context.

**Comment by thvdveld** [comment]
I'm not very familiar with DNS, so I don't feel comfortable reviewing this. But I'll take a look at it tomorrow!

Also, a change slipped into master that messed with 6LoWPAN fragmentation, the fix is in #681.

**Comment by benbrittain** [comment]
oh good! I'm glad you told me that because I literally just started a git bisect trying to figure out what broke :)

File: src/socket/dns.rs [hunk 2]
```diff
@@ -200,6 +211,12 @@ impl<'a> Socket<'a> {
 
         let mut raw_name: Vec<u8, MAX_NAME_LEN> = Vec::new();
 
+        let mut mdns = false;
+        if name.split(|&c| c == b'.').last().unwrap() == b"local" {
```
Comment (thvdveld) [inline src/socket/dns.rs:223]: What would happen if the name is just "local", without a dot. Then `.last()` will return "local". Is this correct behaviour? Like I said, I don't know anything about mDNS, so it could be that this is valid.

File: src/socket/dns.rs [hunk 3]
```diff
@@ -81,6 +90,7 @@ struct PendingQuery {
     delay: Duration,
 
     server_idx: usize,
+    mdns: bool,
```
Comment (Dirbaio) [inline src/socket/dns.rs]: This could be a "Protocol" enum, with valus "Dns" and "Mdns".

**Comment by Dirbaio** [comment]
Looks good to me overall. Could you add a Cargo feature for MDNS, so that binaries not using it don't get increased code size?

File: src/socket/dns.rs (hunk 2 above)
Comment (benbrittain) [reply src/socket/dns.rs:223]: My understanding is that "local" without anything else refers to the top-level domain, in which case the RFC does say under "3.0 Multicast DNS Names" that this has the special semantics.

File: src/socket/dns.rs (hunk 3 above)
Comment (benbrittain) [reply src/socket/dns.rs]: I don't think the API should encourage thinking about mDNS and DNS queries as fundamentally different protocols

**Comment by benbrittain** [comment]
I don't think this will have much meaningful code size, but I created a feature. I think future mDNS service discovery functionality should also go under that feature set.

I left the mdns enum inside the `PendingQuery` struct intentionally. with "socket-mdns" disabled, it should be trivially optimized out and it keeps the cfg flow less confusing below. Happy to make the removal from the PendingQuery struct explicit if you'd prefer though.

File: src/socket/dns.rs (hunk 3 above)
Comment (benbrittain) [reply src/socket/dns.rs]: I did something in this vein when I put this behind a feature flag

File: src/iface/interface.rs [hunk 4]
```diff
@@ -1015,22 +1015,14 @@ impl<'a> Interface<'a> {
         self.inner.now = timestamp;
 
         #[cfg(feature = "proto-ipv4-fragmentation")]
-        if let Err(e) = self
-            .fragments
+        self.fragments
```
Comment (Dirbaio) [inline src/iface/interface.rs:1018]: seems there are some unrelated changes

Comment (Dirbaio) [reply src/iface/interface.rs:1018]: ah it's clippy lints, I see

**Review by Dirbaio — APPROVED** [review]
Thank you!

bors r+

(Yes, the enum becoming single-case is fine, it's the clearest way to do this. We do this in other places already, for example for ipv4/v6)
